src/math_functions.cpp

```diff
@@ -0,0 +1,818 @@
+          ((((d[0]*q + d[1])*q + d[2])*q + d[3])*q + 1.0);
+  }
+
+  // Refinement based on Newton's method
+
+  z = z - (0.5 * std::erfc(-z / std::sqrt(2.0)) - p) * std::sqrt(2.0 * PI) *
+       std::exp(0.5 * z * z);
+
+  return z;
+
+}
+
+//==============================================================================
+// T_PERCENTILE calculates the percentile of the Student's t distribution with
+// a specified probability level and number of degrees of freedom
+//==============================================================================
+
+double __attribute__ ((const)) t_percentile_c(const double p, const int df){
+  double t;
+  double n;
+  double k;
+  double z;
+  double z2;
+
+  if (df == 1) {
+    // For one degree of freedom, the t-distribution becomes a Cauchy
+    // distribution whose cdf we can invert directly
+
+    t = std::tan(PI*(p - 0.5));
+  } else if (df == 2) {
+     // For two degrees of freedom, the cdf is given by 1/2 + x/(2*sqrt(x^2 +
+     // 2)). This can be directly inverted to yield the solution below
+
+    t = 2.0 * std::sqrt(2.0)*(p - 0.5) /
+         std::sqrt(1. - 4. * std::pow(p - 0.5, 2.));
+  } else {
+    // This approximation is from E. Olusegun George and Meenakshi Sivaram, "A
+    // modification of the Fisher-Cornish approximation for the student t
+    // percentiles," Communication in Statistics - Simulation and Computation,
+    // 16 (4), pp. 1123-1132 (1987).
+
+    n = static_cast<double>(df);
+    k = 1. / (n - 2.);
+    z = normal_percentile_c(p);
+    z2 = z * z;
+    t = std::sqrt(n * k) * (z + (z2 - 3.) * z * k / 4. + ((5. * z2 - 56.) * z2 +
+         75.) * z * k * k / 96. + (((z2 - 27.) * 3. * z2 + 417.) * z2 - 315.) *
+         z * k * k * k / 384.);
+  }
+
+  return t;
+}
+
+//==============================================================================
+// CALC_PN calculates the n-th order Legendre polynomials at the value of x.
+//==============================================================================
+
+void calc_pn_c(const int n, const double x, double pnx[]) {
+  int l;
+
+  pnx[0] = 1.;
+  if (n >= 1) {
+    pnx[1] = x;
+  }
+
+  // Use recursion relation to build the higher orders
+  for (l = 1; l < n; l ++) {
+    pnx[l + 1] = (static_cast<double>(2 * l + 1) * x * pnx[l] -
+                  static_cast<double>(l) * pnx[l - 1]) /
+         (static_cast<double>(l + 1));
+  }
+}
+
+//==============================================================================
+// EVALUATE_LEGENDRE Find the value of f(x) given a set of Legendre coefficients
+// and the value of x
+//==============================================================================
+
+double __attribute__ ((const)) evaluate_legendre_c(const int n,
+                                                   const double data[],
+                                                   const double x) {
+  double val;
+  int l;
+  double* pnx = new double[n + 1];
```

Comment:
No need for dynamic memory here.  Just use `double pnx[n+1]` and get rid of the `delete[] pnx` below.

Reply:
- [x]